earthpy/spatial.py

```diff
@@ -370,6 +370,107 @@ def crop_image(raster, geoms, all_touched=True):
     return out_image, out_meta
 
 
+def crop_all(
+    band_paths,
+    output_dir,
+    geoms,
+    overwrite=False,
+    all_touched=True,
+    verbose=True,
+):
+    """Takes a list of rasters and a boundary, and crops them efficiently.
+
+    Parameters
+    ----------
+    band_paths : list of file paths
+        List of paths of bands that are to be cropped.
+    output_dir : list or string
+        If a list is provided, then the output files from this function will be written to
+        each path in the list. This allows for customization of file names.
+        If a string is provided, then the output files from the function will be written
+        to that directory with the name of the original file plus a suffix of "_crop".
+    geoms : geopandas geodataframe or list of polygons
+        The spatial polygon boundaries in GeoJSON-like dict format
+        to be used to crop the image. All data outside of the polygon
+        boundaries will be set to nodata and/or removed from the image.
+    overwrite : bool (default=False)
+        Disallows files to be overwritten if they exist already. Can be changed so that files
+        can be overwritten with each run of the function.
+    all_touched : bool (default=True)
+        Include a pixel in the mask if it touches any of the
+        shapes. If False, include a pixel only if its center is within one of
+        the shapes, or if it is selected by Bresenham's line algorithm.
+        (from rasterio)
+    verbose : bool (default=True)
+        Returns a list of file paths with the files that were created in the function. Can
+        be turned off so the user doesn't have to assign the function to a variable.
+
+    Returns
+    ----------
+    return files : list
+        List of the files created by the function.
+
+    Example
+    -------
+        >>> import os
+        >>> import earthpy.spatial as es
+        >>> import geopandas as gpd
+        >>> from earthpy.io import path_to_example
+        >>> band_fnames = ["red.tif", "green.tif", "blue.tif"]
+        >>> band_paths = [path_to_example(fname) for fname in band_fnames]
+        >>> rmnp = gpd.read_file(path_to_example("rmnp.shp"))
+        >>> output_dir = os.path.commonpath(band_paths)
+        >>> output_files = es.crop_all(band_paths, output_dir, rmnp, overwrite=True)
+        >>> len(output_files)
+        3
+        >>> os.path.isfile(output_files[0])
+        True
+        >>> # Cleaning up example data
+        >>> for bands in output_files:
+        ...     os.remove(bands)
+
+    """
+    try:
+        if not os.path.exists(os.path.commonpath(output_dir)):
+            raise TypeError(
+                "The output directory that you provided does not exist"
+            )
+        list_type = True
+        if len(band_paths) != len(output_dir):
+            raise TypeError(
+                "The list of input bands does not match the length of the list of output file names."
+            )
+    except ValueError:
```

Comment:
@nkorinek it looks like this line isn't tested properly
https://codecov.io/gh/earthlab/earthpy/pull/342/diff?src=pr&el=tree#D4-446 please have a look at it. also it looks like you are raising a value error and then a type error . any reason for that. output directory not existing seems like a value error not a type issue but please have a closer look into this.

Reply:
Hey! So I'm raising a TypeError here instead of a ValueError because of how I'm testing to see if the user passed in a string or a list. If the user passes in a string, `os.path.commonpath()` raises a ValueError. This is a way to test for a list without using `if type(output_dir) == list`. Because of this, if the bad directories raise a ValueError, it will go into the `except` block when it wasn't supposed to, since it's the same error raised if a string is passed into `os.path.commonpath()`. If you're not a fan of this, and want me to just make them ValueErrors and get rid of the try/except blocks, I can do that too, would be easy. @lwasser

Reply:
ok... @nkorinek i finally understand what you are trying to do. here is where i *thiknk* it might fail. but please do let me know if i'm still not fully understanding. 
This: passes because the dir Users/leahwasser exists even tho documents2 does NOT exist. 
so if a user wanted paths in different directories this will fail. i'm starting to think that this option will be confusing for a user -- the option of allowing a list of paths and them placing things in different directories because it assumes that the images that they provide are in the correct order as the paths. it just seems complicated to me to test and also for the user. I wonder if we want to consider something simpler. i may start a little slack convo with max and joe and you!! 

`os.path.commonpath(["/Users/leahwasser/Documents/test.tif", "/Users/leahwasser/Documents2/test.tif"])`

Reply:
we can resolve this is you modify the code to take a simpler approach!